src/Microsoft.AspNetCore.SignalR.Client.Core/HubConnection.cs

```diff
@@ -1129,61 +1129,54 @@ public ConnectionState(ConnectionContext connection, HubConnection hubConnection
 
             public void AddInvocation(InvocationRequest irq)
             {
-                lock (_lock)
+                 
+                    
+                if (!_pendingCalls.TryAdd(irq.InvocationId, irq))
                 {
-                    if (_pendingCalls.ContainsKey(irq.InvocationId))
-                    {
-                        Log.InvocationAlreadyInUse(_hubConnection._logger, irq.InvocationId);
-                        throw new InvalidOperationException($"Invocation ID '{irq.InvocationId}' is already in use.");
-                    }
-                    else
-                    {
-                        _pendingCalls.Add(irq.InvocationId, irq);
-                    }
+                    Log.InvocationAlreadyInUse(_hubConnection._logger, irq.InvocationId);
+                    throw new InvalidOperationException($"Invocation ID '{irq.InvocationId}' is already in use.");
                 }
+                     
+                 
             }
 
             public bool TryGetInvocation(string invocationId, out InvocationRequest irq)
             {
-                lock (_lock)
-                {
+                 
                     return _pendingCalls.TryGetValue(invocationId, out irq);
-                }
+                 
             }
 
             public bool TryRemoveInvocation(string invocationId, out InvocationRequest irq)
             {
-                lock (_lock)
-                {
-                    if (_pendingCalls.TryGetValue(invocationId, out irq))
-                    {
-                        _pendingCalls.Remove(invocationId);
-                        return true;
-                    }
-                    else
-                    {
-                        return false;
-                    }
-                }
+                
+
+                    return  _pendingCalls.TryRemove(invocationId, out irq);
+                     
+                
             }
 
             public void CancelOutstandingInvocations(Exception exception)
             {
                 Log.CancelingOutstandingInvocations(_hubConnection._logger);
 
-                lock (_lock)
+
+                //https://msdn.microsoft.com/en-us/library/dd287131.aspx
+                // From the documentation:
+                //The enumerator returned from the dictionary is safe to use concurrently with reads and writes to the dictionary,
+                //however it does not represent a moment-in-time snapshot of the dictionary.
+                //The contents exposed through the enumerator may contain modifications made to the dictionary after GetEnumerator was called.
+                foreach (var outstandingCallKeyPair in _pendingCalls)
```

Comment:
Is there a reason that you iterate through the key pairs instead of just the values of `_pendingCalls`. It looks like you access the `Value` field in every reference to `outstandingCallKeyPair`.

Reply:
The only reason to access the keypair instead of the collection of values is because the documentation doesn't give any guarantees about the Values thread safety  Enumerator being return. They only  give assurance about the concurrent Dictionary enumerator  being return. 
If you use the values enumerator, that is a not threadsafe copy of the state of those values, that may not represent the reallity of that dictionary value objects wich can be a problem if you call Dispose over an object that may  already  be removed from that dictionary.